### PR TITLE
Make in-place FFT optional

### PR DESCRIPTION
**Pull request by samhatfield**
It seems that hipFFT has a problem with in-place FFTs. Running on LUMI-G (MI250x), this error is thrown at runtime:

```
GPU runtime error at 1
GPU runtime error in file '/pfs/lustrep3/scratch/project_465000454/hatfield/jube-suites/ectrans/ectrans_gpu/000005/000000_clone/work/ectrans-bundle/source/ectrans/src/trans/gpu/algor/hicfft.hip.cpp'
GPU runtime error at 2
GPU runtime error line '48'
GPU runtime error at 3
GPU runtime error 6: HIPFFT_EXEC_FAILED
terminating!
ABOR1     [PROC=1,THRD=1] from [/pfs/lustrep3/scratch/project_465000454/hatfield/jube-suites/ectrans/ectrans_gpu/000005/000000_clone/work/ectrans-bundle/source/ectrans/src/trans/gpu/algor/hicfft.h +48] : Error in FFT
```

Some investigations with @PaulMullowney showed that if you create a separate buffer for the output, the error vanishes.

With this change we are now finally able to run the optimised GPU version of ecTrans (now _the_ GPU version of ecTrans) on AMD GPUs, at least with a single MI250x GPU on LUMI-G. Multi-GPU runs are still a work in progress.

This PR makes in-place FFT compile-time adjustable. For now, in-place is enabled for cuFFT, but disabled for hipFFT.

@PaulMullowney and I will see what we can do to flag this to hipFFT developers so we can eventually remove this option.

Note that earlier `HIPFFT_PARSE_ERROR`s no longer appear with ROCm 6.0.2, which is the new default for LUMI-G.

Note that this requires testing on an Nvidia GPU. Hopefully we have not ruined things there.

This PR builds on https://github.com/ecmwf-ifs/ectrans/pull/150 so best to merge that one first.

### CODE REVIEW

**Comment by samhatfield** [comment]
@lukasm91 can you see this having any performance implications for CUDA/cuFFT execution? I think I've put all the expensive stuff inside the guards so it shouldn't be compiled on an Nvidia platform.

File: src/trans/gpu/internal/ftdir_mod.F90 [hunk 1]
```diff
@@ -77,7 +94,12 @@ SUBROUTINE FTDIR(ALLOCATOR,HFTDIR,PREEL_REAL,PREEL_COMPLEX,KFIELD)
 
     INTEGER(KIND=JPIM) :: KGL
 
+#ifdef IN_PLACE_FFT
     PREEL_COMPLEX => PREEL_REAL
+#else
+    CALL ASSIGN_PTR(PREEL_COMPLEX, GET_ALLOCATION(ALLOCATOR, HFTDIR%HREEL_COMPLEX),&
+      & 1_C_SIZE_T, int(KFIELD*D%NLENGTF*SIZEOF(PREEL_COMPLEX(1)),kind=c_size_t))
```
Comment (lukasm91) [inline src/trans/gpu/internal/ftdir_mod.F90:101]:
```suggestion
      & 1_C_SIZE_T, INT(KFIELD*D%NLENGTF*SIZEOF(PREEL_COMPLEX(1)),KIND=CSIZE_T))
```
I know we are not very consistent everywhere, but I suggest not add more caps inconsistencies.

File: src/trans/gpu/internal/ftinv_mod.F90 [hunk 2]
```diff
@@ -10,20 +10,33 @@
...
     IMPLICIT NONE
 
     TYPE(BUFFERED_ALLOCATOR), INTENT(INOUT) :: ALLOCATOR
+    INTEGER(KIND=JPIM), INTENT(IN) :: KF_FS
     TYPE(FTINV_HANDLE) :: HFTINV
+
+    REAL(KIND=JPRBT) :: DUMMY
+
+#ifndef IN_PLACE_FFT
+    HFTINV%HREEL_REAL = RESERVE(ALLOCATOR, int(D%NLENGTF*KF_FS*SIZEOF(DUMMY),kind=c_size_t))
```
Comment (lukasm91) [inline src/trans/gpu/internal/ftinv_mod.F90:38]:
```suggestion
    HFTINV%HREEL_REAL = RESERVE(ALLOCATOR, INT(D%NLENGTF*KF_FS*SIZEOF(DUMMY),KIND=C_SIZE_T))
```

File: src/trans/gpu/internal/ftinv_mod.F90 [hunk 3]
```diff
@@ -77,7 +92,12 @@ SUBROUTINE FTINV(ALLOCATOR,HFTINV,PREEL_COMPLEX,PREEL_REAL,KFIELD)
 
     INTEGER(KIND=JPIM) :: KGL
 
+#ifdef IN_PLACE_FFT
     PREEL_REAL => PREEL_COMPLEX
+#else
+    CALL ASSIGN_PTR(PREEL_REAL, GET_ALLOCATION(ALLOCATOR, HFTINV%HREEL_REAL),&
+      & 1_C_SIZE_T, int(KFIELD*D%NLENGTF*SIZEOF(PREEL_REAL(1)),kind=c_size_t))
```
Comment (lukasm91) [inline src/trans/gpu/internal/ftinv_mod.F90:99]:
```suggestion
      & 1_C_SIZE_T, INT(KFIELD*D%NLENGTF*SIZEOF(PREEL_REAL(1)),KIND=C_SIZE_T))
```

**Review by lukasm91 — COMMENTED** [review]
Yes, this looks like it should not have perf implications (and I also quickly checked it on our machines)

File: src/trans/gpu/internal/ftdir_mod.F90 [hunk 4]
```diff
@@ -10,19 +10,35 @@
...
+
+    TYPE(BUFFERED_ALLOCATOR), INTENT(INOUT) :: ALLOCATOR
+    INTEGER(KIND=JPIM), INTENT(IN) :: KF_FS
     TYPE(FTDIR_HANDLE) :: HFTDIR
-  END FUNCTION
+
+    REAL(KIND=JPRBT) :: DUMMY
+
+#ifndef IN_PLACE_FFT
+    HFTDIR%HREEL_COMPLEX = RESERVE(ALLOCATOR, INT(KF_FS*D%NLENGTF*SIZEOF(DUMMY), KIND=C_SIZE_T))
```
Comment (wdeconinck) [inline src/trans/gpu/internal/ftdir_mod.F90:39]: `SIZEOF` is a non-standard extension. Some compilers could complain.
Standard in F2008 is `STORAGE_SIZE`, which gives you `bits` (not bytes!) and `C_SIZEOF`, which gives you `bytes`.

File: src/trans/gpu/internal/ftdir_mod.F90 [hunk 5]
```diff
@@ -77,7 +94,12 @@ SUBROUTINE FTDIR(ALLOCATOR,HFTDIR,PREEL_REAL,PREEL_COMPLEX,KFIELD)
 
     INTEGER(KIND=JPIM) :: KGL
 
+#ifdef IN_PLACE_FFT
     PREEL_COMPLEX => PREEL_REAL
+#else
+    CALL ASSIGN_PTR(PREEL_COMPLEX, GET_ALLOCATION(ALLOCATOR, HFTDIR%HREEL_COMPLEX),&
+      & 1_C_SIZE_T, INT(KFIELD*D%NLENGTF*SIZEOF(PREEL_COMPLEX(1)),KIND=C_SIZE_T))
```
Comment (wdeconinck) [inline src/trans/gpu/internal/ftdir_mod.F90:101]: also here

File: src/trans/gpu/internal/ftinv_mod.F90 [hunk 6]
```diff
@@ -10,20 +10,33 @@
...
     IMPLICIT NONE
 
     TYPE(BUFFERED_ALLOCATOR), INTENT(INOUT) :: ALLOCATOR
+    INTEGER(KIND=JPIM), INTENT(IN) :: KF_FS
     TYPE(FTINV_HANDLE) :: HFTINV
+
+    REAL(KIND=JPRBT) :: DUMMY
+
+#ifndef IN_PLACE_FFT
+    HFTINV%HREEL_REAL = RESERVE(ALLOCATOR, INT(D%NLENGTF*KF_FS*SIZEOF(DUMMY),KIND=C_SIZE_T))
```
Comment (wdeconinck) [inline src/trans/gpu/internal/ftinv_mod.F90:38]: and here

File: src/trans/gpu/internal/ftinv_mod.F90 [hunk 7]
```diff
@@ -77,7 +92,12 @@ SUBROUTINE FTINV(ALLOCATOR,HFTINV,PREEL_COMPLEX,PREEL_REAL,KFIELD)
 
     INTEGER(KIND=JPIM) :: KGL
 
+#ifdef IN_PLACE_FFT
     PREEL_REAL => PREEL_COMPLEX
+#else
+    CALL ASSIGN_PTR(PREEL_REAL, GET_ALLOCATION(ALLOCATOR, HFTINV%HREEL_REAL),&
+      & 1_C_SIZE_T, INT(KFIELD*D%NLENGTF*SIZEOF(PREEL_REAL(1)),KIND=C_SIZE_T))
```
Comment (wdeconinck) [inline src/trans/gpu/internal/ftinv_mod.F90:99]: and here..
Perhaps there are more instances in the code. This seems like a typical copy-paste-edit line.

**Review by wdeconinck — APPROVED** [review]

File: src/trans/gpu/internal/ftdir_mod.F90 (hunk 4 above)
Comment (samhatfield) [reply src/trans/gpu/internal/ftdir_mod.F90:39]: `SIZEOF` occurs in a few places in the GPU tree. @lukasm91 any strong feelings about switching to `8*STORAGE_SIZE(DUMMY)`?

**Pull request merged**